File: src/versioning/PijulBackend.js
const { runCommand } = require('../util');
const VersioningBackend = require('./VersioningBackend');

class PijulBackend extends VersioningBackend {
  async add(file) {
    await runCommand('pijul', ['add', file]);
  }

  async record(message) {
    await runCommand('pijul', ['record', '-m', message]);
  }

  async unrecord(hash) {
    await runCommand('pijul', ['unrecord', hash]);
  }

  async diff() {
    const { stdout } = await runCommand('pijul', ['diff']);
    return stdout;
  }

  async status() {
    const { stdout: channel } = await runCommand('pijul', ['channel']);
    const currentChannel = channel.split('\n').find(line => line.startsWith('*')).substring(1).trim();
    const { stdout: status } = await runCommand('pijul', ['status', '--porcelain']);
    const files = status.split('\n').filter(line => line.trim() !== '').map(line => {
        const parts = line.trim().split(' ');
        return {
            status: parts[0],
            file: parts.slice(1).join(' '),
        };
    });
    return JSON.stringify({
        channel: currentChannel,
        files,
    }, null, 2);
  }

  async channel(subcommand, name) {
    if (subcommand === 'new') {
      await runCommand('pijul', ['channel', 'new', name]);
    } else if (subcommand === 'switch') {
      await runCommand('pijul', ['channel', 'switch', name]);
    } else if (subcommand === 'list') {
      const { stdout } = await runCommand('pijul', ['channel']);
      return stdout;
    }
  }

  async patch(subcommand, name) {
    if (subcommand === 'add') {
      await runCommand('pijul', ['patch', 'add', name]);
    } else if (subcommand === 'list') {
      const { stdout } = await runCommand('pijul', ['log']);
      return stdout;
    }
  }

  async apply(patch) {
    await runCommand('pijul', ['apply', patch]);
  }

  async conflicts() {
    const { stdout } = await runCommand('pijul', ['credit']);
    const conflicts = stdout
      .split('\n')
      .filter((line) => line.startsWith('C'))
      .map((line) => {
        const parts = line.split(' ');
        return {
          hash: parts[1],
          author: parts[2],
          date: parts[3],
        };
      });
    return JSON.stringify(conflicts, null, 2);
  }

  async revert(file) {
    await runCommand('pijul', ['reset', file]);
  }

  async revertAll() {
    await runCommand('pijul', ['reset']);
  }

  async undo() {
    const { stdout } = await runCommand('pijul', ['log', '--limit', '1', '--hash-only']);
    const hash = stdout.trim();
    if (hash) {
      await this.unrecord(hash);
    }
  }
}

module.exports = PijulBackend;

File: java/src/main/java/com/pijul/aider/PijulBackend.java
package com.pijul.aider;

import java.util.List;
import java.util.ArrayList;
import com.pijul.aider.versioning.VersioningBackend;
import java.util.concurrent.CompletableFuture;
import java.io.IOException;
import java.io.InputStream;
import java.io.ByteArrayOutputStream;

public class PijulBackend extends VersioningBackend implements Backend {
    @Override
    public CompletableFuture<Void> add(String file) {
        return executeCommand("pijul", "add", file);
    }

    @Override
    public CompletableFuture<Void> unstage(String file) {
        return executeCommand("pijul", "remove", file);
    }

    @Override
    public CompletableFuture<Void> record(String message) {
        return executeCommand("pijul", "record", "-m", message);
    }

    @Override
    public CompletableFuture<Void> unrecord(String hash) {
        return executeCommand("pijul", "unrecord", hash);
    }

    @Override
    public CompletableFuture<Void> channel(String subcommand, String name) {
        return executeCommand("pijul", "channel", subcommand, name);
    }

    @Override
    public CompletableFuture<Void> patch(String subcommand, String name) {
        return executeCommand("pijul", "patch", subcommand, name);
    }

    @Override
    public CompletableFuture<Void> apply(String patch) {
        return executeCommand("pijul", "apply", patch);
    }

    @Override
    public CompletableFuture<String> conflicts() {
        return executeCommandWithOutput("pijul", "conflicts");
    }

    @Override
    public CompletableFuture<Void> revert(String file) {
        return executeCommand("pijul", "revert", file);
    }

    @Override
    public CompletableFuture<Void> revertAll() {
        return executeCommand("pijul", "revert", ".");
    }

    @Override
    public CompletableFuture<String> diff() {
        return executeCommandWithOutput("pijul", "diff");
    }

    @Override
    public CompletableFuture<String> status() {
        return executeCommandWithOutput("pijul", "status");
    }

    @Override
    public CompletableFuture<Void> clear() {
        return executeCommand("pijul", "clear");
    }

    @Override
    public CompletableFuture<List<String>> listTrackedFiles() {
        return executeCommandWithListOutput("pijul", "ls");
    }

    @Override
    public CompletableFuture<List<String>> listUntrackedFiles() {
        return executeCommandWithListOutput("pijul", "ls", "--untracked");
    }

    @Override
    public CompletableFuture<Void> initialize() {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletableFuture<Void> shutdown() {
        return CompletableFuture.completedFuture(null);
    }

    @Override
    public CompletableFuture<Void> undo() {
        return executeCommand("pijul", "undo");
    }

    private CompletableFuture<Void> executeCommand(String... commands) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                ProcessBuilder pb = new ProcessBuilder(commands)
                    .redirectErrorStream(true);
                Process process = pb.start();
                int exitCode = process.waitFor();
                if (exitCode != 0) {
                    throw new RuntimeException("Command failed: " + String.join(" ", commands));
                }
                return null;
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        }).thenApply(result -> null);
    }

    private CompletableFuture<String> executeCommandWithOutput(String... commands) {
        return CompletableFuture.supplyAsync(() -> {
            try {
                ProcessBuilder pb = new ProcessBuilder(commands)
                    .redirectErrorStream(true);
                Process process = pb.start();
                
                // Read output
                InputStream inputStream = process.getInputStream();
                ByteArrayOutputStream result = new ByteArrayOutputStream();
                byte[] buffer = new byte[1024];
                int bytesRead;
                while ((bytesRead = inputStream.read(buffer)) != -1) {
                    result.write(buffer, 0, bytesRead);
                }
                String output = result.toString().trim();
                
                int exitCode = process.waitFor();
                if (exitCode != 0) {
                    throw new RuntimeException("Command failed: " + String.join(" ", commands));
                }
                return output;
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
        });
    }

    private CompletableFuture<List<String>> executeCommandWithListOutput(String... commands) {
        return executeCommandWithOutput(commands).thenApply(output -> {
            return java.util.Arrays.asList(output.split("\\s+"));
        });
    }
}